### ISSUE TITLE
Questions about personal-copilot blog

### ISSUE THREAD

**Issue by cmosguy** [huggingface/blog#1619]
Hey @pacman100 

Your blog here was awesome and I learned a ton!  

[personal-copilot.md](https://github.com/huggingface/blog/blob/cb6683e494beca3b04d2950ae99d3b35d6634fbf/personal-copilot.md)

This is so helpful for us folks trying to implement our very own code co-pilots on premise.

I had a question of questions:

1.  You showed the impact of HumanEval.  I am assuming you ran the big code evaluation harness: [bigcode-project/bigcode-evaluation-harness](https://github.com/bigcode-project/bigcode-evaluation-harness) .  My questions is how do you reason about and evaluate the code completions of what you finetuned on.  For example, you are you saw `pass@1 = 33.37`.  Now, this checks for catastrophic forgetting. However, now how can you add in and evaluate your new fine-tuned data from the HF repos you trained on to check that your training is working properly.  I know you are interacting with the code as a human and in the pictures you included in your blog, but how do you do this **quantitatively** without manually writing `humaneval` tests?
2. For the chat/qa component, is there an extension for VS code that allows you to interact with that as well?
3. I noticed you went through a lot of hoops to do [pacman100/mlc-llm](https://github.com/pacman100/mlc-llm).  Was there a reason you did this instead of just deploying a docker image locally of the [huggingface/text-generation-inference](https://github.com/huggingface/text-generation-inference) instead?

Thank you!

**Comment by pcuenca** [comment]
cc @sayakpaul too :)

**Comment by sayakpaul** [comment]
> I noticed you went through a lot of hoops to do [pacman100/mlc-llm](https://github.com/pacman100/mlc-llm). Was there a reason you did this instead of just deploying a docker image locally of the [huggingface/text-generation-inference](https://github.com/huggingface/text-generation-inference) instead?

I am guess this was more of a personal choice. 

> For the chat/qa component, is there an extension for VS code that allows you to interact with that as well?

We didn't. In an IDE, infilling is the most use case and we focused on those.

**Comment by cmosguy** [comment]
@sayakpaul Thanks for answering my questions. If you don't mind me asking one more. 

When I try to run your FDSP example with accelerate on 4x A100 80GB. I keep getting cuda out of memory errors. Do you have any recommendations why this could be happening? This error happens when training starts. Oh btw, I cannot tell what versions of transformers, left, accelerate, etc you used. I feel like if I sync to the main line of these repos a lot of errors are getting introduced. Do you recommend anything?

**Comment by sayakpaul** [comment]
For that, I welcome @pacman100 to answer it.

**Comment by pacman100** [comment]
Hello @cmosguy, thank you for reading the blog post and happy that you found it insightful 🤗!

With respect to the questions:

> You showed the impact of HumanEval. I am assuming you ran the big code evaluation harness: [bigcode-project/bigcode-evaluation-harness](https://github.com/bigcode-project/bigcode-evaluation-harness) . My questions is how do you reason about and evaluate the code completions of what you finetuned on. For example, you are you saw pass@1 = 33.37. Now, this checks for catastrophic forgetting. However, now how can you add in and evaluate your new fine-tuned data from the HF repos you trained on to check that your training is working properly. I know you are interacting with the code as a human and in the pictures you included in your blog, but how do you do this **quantitatively** without manually writing humaneval tests?

This is definitely a major challenge. Evaluating quantitatively is a topic that I pondered upon and discussed with colleagues while working on this. We came up with the following evaluation strategies that were out of scope of this work:
1. The best evaluation process would be to perform A/B testing on end users/developers. This is direct human feedback and reflects the usefulness of the model in a true sense. The metrics to track would be the percentage of suggestions accepted (acceptance rate), reduction in coding iteration to PR/build/test, fraction of code added by model,  avg characters added per acceptance [1].
2. Have an automated system to create test datasets with single/multi-line masks. Measure the time taken by model to fill the masks, run it through the tests for the project and measure the percentage of compilable completions and percentage of completions passing the test suite. 
3. Surveys like below [2] will give a more holistic developer productivity assessments. It is important to note that **"developer productivity cannot be reduced to a single dimension (or metric!)"**[3] :
![image](https://github.com/huggingface/blog/assets/13534540/5c68aca6-feb6-4611-b6a2-fc07a25777b8)

[1] https://blog.research.google/2022/07/ml-enhanced-code-completion-improves.html
[2] https://github.blog/2022-09-07-research-quantifying-github-copilots-impact-on-developer-productivity-and-happiness/
[3] https://queue.acm.org/detail.cfm?id=3454124

Additional good reads:
[4] https://github.blog/2022-07-14-research-how-github-copilot-helps-improve-developer-productivity/
[5] https://github.com/features/preview/copilot-x

cc @lvwerra and @loubnabnl for their thoughts around this important aspect of measuring the usefulness of code completion models for improving the developer productivity.

> For the chat/qa component, is there an extension for VS code that allows you to interact with that as well?

I use Gradio apps for this purpose. However, I agree that it would be great to have it as part of the IDE to reduce the context-switching time and be in the flow. I found these paid extensions [CodeGPT](https://x.com/dani_avila7/status/1720822445270085785?s=20) and the [Copilot-X](https://github.com/features/preview/copilot-x) to support it.

> I noticed you went through a lot of hoops to do [pacman100/mlc-llm](https://github.com/pacman100/mlc-llm). Was there a reason you did this instead of just deploying a docker image locally of the [huggingface/text-generation-inference](https://github.com/huggingface/text-generation-inference) instead?

I might be wrong, but I think TGI only supports Nvidia GPUs at the moment (cc @OlivierDehaene for confirmation). I wanted to run the model locally on my Mac M1 pro GPU and hence using the MLC which provides great performance and is user friendly. The blogpost also shows how to use HF Inference endpoints to host the model in this section [How do I use it in VS Code?](https://github.com/huggingface/blog/blob/cb6683e494beca3b04d2950ae99d3b35d6634fbf/personal-copilot.md#how-do-i-use-it-in-vs-code)

**Comment by OlivierDehaene** [comment]
Well it could be great to have a PR to add support for the `mps` device but you are right that today TGI would be pretty slow on M1 as it would fallback to `cpu` device.

**Comment by pacman100** [comment]
> When I try to run your FDSP example with accelerate on 4x A100 80GB. I keep getting cuda out of memory errors. Do you have any recommendations why this could be happening? This error happens when training starts. Oh btw, I cannot tell what versions of transformers, peft, accelerate, etc you used. I feel like if I sync to the main line of these repos from today, a lot of errors are getting introduced. Do you recommend anything?

Using the latest releases of transformers, Accelerate and PEFT, I can run FSDP example on 4xA100 80GB. However, it almost runs on the edge consuming 80GB out of the 81.920GB.  For time-being, I notice that using the transformers FA2 support leads to huge loss. So make the below changes in `DHS-LLM-Workshop/personal_copilot/training/train.py`:
```
model = AutoModelForCausalLM.from_pretrained(
        args.model_path,
        load_in_8bit=load_in_8bit,
        quantization_config=bnb_config,
        device_map=device_map,
        use_cache=not args.no_gradient_checkpointing,
        trust_remote_code=True,
#         use_flash_attention_2=args.use_flash_attn,  
    )

...

def main(args):
    if args.use_flash_attn:
        from starcoder_flash_attn_monkey_patch import replace_starcoder_attn_with_flash_attn
        replace_starcoder_attn_with_flash_attn()
```

cc @younesbelkada wrt undesired behaviour by FA2 flag of transformers.

**Comment by cmosguy** [comment]
@pacman100 thank you for all the thoughtful answers.  I've learned a ton by replicating what you are doing there at 🤗.  By replicating your examples I was able to implement a finetuned model on our code base.  Also a shoutout to @loubnabnl  for making amazing Santacoder examples with FIM support was a huge help to get started.  Seriously, you guys are making a huge impact in the community, really appreciate it!
